compat/src/suspense.js

```diff
@@ -61,40 +53,32 @@ Suspense.prototype = new Component();
 Suspense.prototype._childDidSuspend = function(promise) {
 	/** @type {import('./internal').SuspenseComponent} */
 	const c = this;
-	c._suspensions.push(promise);
 
 	const onSuspensionComplete = () => {
-		// From https://twitter.com/Rich_Harris/status/1125850391155965952
-		c._suspensions[c._suspensions.indexOf(promise)] =
-			c._suspensions[c._suspensions.length - 1];
-		c._suspensions.pop();
-
-		if (c._suspensions.length == 0) {
-			// If fallback is null, don't try to unmount it
-			// `unmount` expects a real VNode, not null values
-			if (c._fallback) {
-				// Unmount current children (should be fallback)
-				unmount(c._fallback);
-			}
-			c._vnode._dom = null;
-
-			c._vnode._children = c.state._parkedChildren;
-			c.setState({ _parkedChildren: null });
+		if (!--c._suspensions) {
+			c._detachOnNextRender = false;
+			c._vnode._children[0]._children = c.state._suspended;
+			c.setState({ _suspended: null });
 		}
 	};
 
-	if (c.state._parkedChildren == null) {
-		c._fallback = c._fallback && cloneElement(c._fallback);
-		c.setState({ _parkedChildren: c._vnode._children });
-		detachDom(c._vnode._children);
-		c._vnode._children = [];
+	if (!c._suspensions++) {
+		c._detachOnNextRender = true;
+		c.setState({ _suspended: c._vnode._children[0]._children });
 	}
-
 	promise.then(onSuspensionComplete, onSuspensionComplete);
 };
 
 Suspense.prototype.render = function(props, state) {
-	return state._parkedChildren ? this._fallback : props.children;
+	if (this._detachOnNextRender) {
+		detachDom(this._vnode._children[0])._children = [];
+		this._detachOnNextRender = false;
+	}
+
+	return [
+		createElement(Fragment, null, state._suspended ? null : props.children),
```

Comment:
Double checking, this doesn't call the unmount lifecycles for children when a component suspends? Probably need to double check that we have a test that covers this.

Reply:
That's actually a very valid point. Pretty sure this will call the unmount lifecycles what is different to react and not really the meaning of suspense

Reply:
There appears to be tests for such scenarios already, which is awesome:
 * [`should not call lifecycle methods when suspending`](https://github.com/preactjs/preact/blob/70195cec93c2cd6a613e2b23f80b0240ca129e21/compat/test/browser/suspense.test.js#L230-L283)
 * [`should call fallback's lifecycle methods when suspending`](https://github.com/preactjs/preact/blob/70195cec93c2cd6a613e2b23f80b0240ca129e21/compat/test/browser/suspense.test.js#L285-L339)
 * [`should keep state of children when suspending`](https://github.com/preactjs/preact/blob/70195cec93c2cd6a613e2b23f80b0240ca129e21/compat/test/browser/suspense.test.js#L341-L388)

The first one (`should not call lifecycle methods when suspending`) tests that the lifecycle hooks of a component with a sibling that suspends. I added a new test that tests the lifecycle hooks of a component that itself suspends. I assumed that the `componentDidMount` shouldn't be called for such a component when it suspends, as React doesn't seem to do so either (see this [Codesandbox example](https://codesandbox.io/s/componentdidmount-not-called-for-suspended-component-wvh66)).

Reply:
But how does that first test succeed? I‘d expect the lifecycles being called with this implementation. Will try to free some time to understand this better 🤔

Reply:
Ah now I see that‘s the detachedClone fiddling you do in render 👍👌